File: Codex/Python/2025/Day7b_Solution_2025.py
def read_grid() -> list[str]:
    for path in ("Day7_Input.txt", "Day7_input.txt"):
        try:
            with open(path, "r", encoding="utf-8") as file:
                return [line.rstrip("\n") for line in file if line.rstrip("\n")]
        except FileNotFoundError:
            continue
    raise FileNotFoundError("Input file not found")


def main() -> None:
    grid = read_grid()

    width = len(grid[0])
    counts = [0] * width
    counts[grid[0].index("S")] = 1

    for row in grid[1:]:
        next_counts = [0] * width

        for col, count in enumerate(counts):
            if count == 0:
                continue

            if row[col] == "^":
                if col > 0:
                    next_counts[col - 1] += count
                if col + 1 < width:
                    next_counts[col + 1] += count
            else:
                next_counts[col] += count

        counts = next_counts

    print(sum(counts))


if __name__ == "__main__":
    main()

File: Codex/Rust/2025/Day7b_Solution_2025.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("Day7_Input.txt")
        .or_else(|_| fs::read_to_string("Day7_input.txt"))
        .expect("failed to read Day7_Input.txt");

    let lines: Vec<&str> = input.lines().collect();
    if lines.is_empty() {
        println!("0");
        return;
    }

    let width = lines[0].len();
    let grid: Vec<&[u8]> = lines.iter().map(|line| line.as_bytes()).collect();

    let start_col = grid[0]
        .iter()
        .position(|&cell| cell == b'S')
        .expect("missing start position");

    let mut current = vec![0_u64; width];
    current[start_col] = 1;

    for row in &grid[1..] {
        let mut next = vec![0_u64; width];

        for col in 0..width {
            let count = current[col];
            if count == 0 {
                continue;
            }

            match row[col] {
                b'^' => {
                    if col > 0 {
                        next[col - 1] += count;
                    }
                    if col + 1 < width {
                        next[col + 1] += count;
                    }
                }
                b'.' | b'S' => {
                    next[col] += count;
                }
                _ => panic!("unexpected cell"),
            }
        }

        current = next;
    }

    let answer: u64 = current.into_iter().sum();
    println!("{answer}");
}
